### PR TITLE
add support for adyen hpp checkout

### PR DESCRIPTION
**Pull request by adeitrick**
Adds support for Adyen HPP Checkout

### CODE REVIEW

File: lib/recurly/adyen.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import Emitter from 'component-emitter';
+import errors from '../errors';
+import {Pricing} from './pricing';
+import {normalize} from '../util/normalize';
```
Comment (chrissrogers) [inline lib/recurly/adyen.js]: Only Emitter is used amongst these imports

File: lib/recurly/adyen.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+const debug = require('debug')('recurly:adyen');
```
Comment (chrissrogers) [inline lib/recurly/adyen.js]: Perhaps add a few debug calls, on construct and start

File: lib/recurly/adyen.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ */
+export function factory (options) {
+  return new Adyen(Object.assign({}, options, { recurly: this }));
+};
+
+
+/**
+ * Initializes an Adyen session.
+ *
+ * @param {Object} options
```
Comment (chrissrogers) [inline lib/recurly/adyen.js]: Can you add the options? Looks like just `@param {Recurly} options.recurly`

File: lib/recurly/adyen.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  constructor (options) {
+    super();
+
+    this._ready = false;
+    this.config = {};
+    this.once('ready', () => this._ready = true);
+    this.recurly = options.recurly
+  }
+
+  start (opts) {
```
Comment (chrissrogers) [inline lib/recurly/adyen.js]: Mind adding a jsdoc entry for the options?

File: lib/recurly/adyen.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      shopperLocale: opts.shopperLocale,
+      currencyCode: opts.currencyCode,
+      skinCode: opts.skinCode
+    };
+
+    const frame = this.recurly.Frame({ height: 600, path: '/adyen/start', payload });
+    frame.once('error', cause => this.error('paypal-tokenize-error', { cause }));
+
+    frame.once('done', function(token) {
+      this.emit('token', token)
```
Comment (chrissrogers) [inline lib/recurly/adyen.js]: I suspect this needs to be in an arrow fn closure to emit properly

```
frame.once('done', token => this.emit('token', token));
```

File: lib/recurly/adyen.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+ * @param {Object} options
+ * @constructor
+ * @public
+ */
+
+class Adyen extends Emitter {
+  constructor (options) {
+    super();
+
+    this._ready = false;
```
Comment (chrissrogers) [inline lib/recurly/adyen.js]: `_ready` is never checked. Should it be on start?

File: lib/recurly/adyen.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    this.config = {};
```
Comment (chrissrogers) [inline lib/recurly/adyen.js]: This isn't used

File: lib/recurly/adyen.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    const payload = {
+      invoiceUuid: opts.invoiceUuid,
+      countryCode: opts.countryCode,
+      shopperLocale: opts.shopperLocale,
+      currencyCode: opts.currencyCode,
+      skinCode: opts.skinCode
+    };
+
+    const frame = this.recurly.Frame({ height: 600, path: '/adyen/start', payload });
+    frame.once('error', cause => this.error('paypal-tokenize-error', { cause }));
```
Comment (chrissrogers) [inline lib/recurly/adyen.js]: Do these errors cover instances of bad options, like an invalid countryCode, shopperLocal, etc. ? Wondering if we need some kind of validation on those values.

Comment (adeitrick) [reply lib/recurly/adyen.js]: Adyen will do its own validations on their end when we try to invoke the modal, although there aren't great errors provided from them around validations. I suppose we could do validations around string length and whatnot here, but not sure how much that will buy us.

File: lib/recurly/adyen.js (hunk 6 above)
Comment (adeitrick) [reply lib/recurly/adyen.js]: not sure its necessary, going to remove

**Pull request merged**